entproto/cmd/protoc-gen-entgrpc/template.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2019-present Facebook
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/hashicorp/go-multierror"
+	"google.golang.org/protobuf/compiler/protogen"
+)
+
+// printTemplate is a utility function to make working with protogen have a more declarative interface.
+// It receives a *protogen.GeneratedFile, a template string with placeholder formatted like "%(variableName)"
+// and a tmplValues map containing the values that should be replaced when the template is rendered.
+//
+// Instead of:
+//	g.P("func New", svcName, "(p string) *", protogen.GoImportPath("context").Ident("Context"))
+// We can use
+//	printTemplate(g, "func New%(svcName)(p string) *%(ctx)", tmplValues{
+//		"svcName": "UserService",
+//		"ctx": protogen.GoImportPath("context").Ident("Context"),
+//	})
+func printTemplate(g *protogen.GeneratedFile, tmpl string, values tmplValues) error {
+	re := regexp.MustCompile(`(%\(.+?\))`)
```

Comment:
Just one comment - unlike Python, the `regexp` package doesn't cache these regular-expressions.
Move this variable to the top-level declarations:

```go
var re = regexp.MustCompile(`(%\(.+?\))`)

func printTemplate(g *protogen.GeneratedFile, tmpl string, values tmplValues) error {
	// ...
}
```